File: MiniGame-Java/src/main/java/rs/lab/mges/game/Mob.java
package rs.lab.mges.game;

import io.github.libsdl4j.api.render.SDL_Texture;
import java.util.Random;
import rs.lab.mges.engine.Drawable.AnimatedSprite;
import rs.lab.mges.engine.SDLUtils.Vector2f;

public abstract class Mob extends AnimatedSprite {

    private static final Random rand = new Random();

    public boolean active;
    public int hp;
    public float speed;
    public int point;

    public Mob(SDL_Texture[] frames, boolean flipped, int hp, float speed, int point) {
        super(frames, 32, 32);

        this.flipped = flipped;
        this.active = true;
        this.hp = hp;
        this.speed = speed;
        this.point = point;

        position.x = flipped ? 640f : -32f;
        velocity = new Vector2f(flipped ? -speed : speed, 0f);
    }

    @Override
    public void update(float dt) {
        if (flipped) {
            if (velocity.x > 0) {
                velocity.x -= 1.0f;
                if (velocity.x < 1.0f) {
                    velocity.x = -speed;
                }
            }
        } else {
            if (velocity.x < 0) {
                velocity.x += 1.0f;
                if (velocity.x >= 0.0f) {
                    velocity.x = speed;
                }
            }
        }

        position.x += velocity.x * dt;

        animTimer += dt;
        if (animTimer > 0.25) {
            animTimer = 0.0f;
            animFrame = 1 - animFrame;
        }

        if (position.x < -32.0f || position.x > 640.0f) {
            active = false;
        }
    }

    public boolean gotHit(Hero player) {
        if (hp-- <= 0) {
            return true;
        }

        velocity.x = player.flipped ? -speed * 1.2f : speed * 1.2f;

        return false;
    }

    public static class MobBee extends Mob {

        private static final int HP = 2;
        private static final float SPEED = 40;
        private static final int POINT = 20;

        public MobBee(RabbitGame game, boolean flipped) {
            super(game.assets.beeFrames, flipped, HP, SPEED, POINT);
            position.y = 280.0f - 32 + rand.nextFloat() * 32.0f;
        }
    }

    public static class MobBlue extends Mob {

        private static final int HP = 3;
        private static final float SPEED = 60;
        private static final int POINT = 50;

        public MobBlue(RabbitGame game, boolean flipped) {
            super(game.assets.blueFrames, flipped, HP, SPEED, POINT);
            position.y = 280.0f - 32 + rand.nextFloat() * 32.0f;
        }
    }

    public static class MobFoxy extends Mob {

        private static final int HP = 2;
        private static final float SPEED = 50;
        private static final int POINT = 100;

        public MobFoxy(RabbitGame game, boolean flipped) {
            super(game.assets.foxyFrames, flipped, HP, SPEED, POINT);
            position.y = 280.0f;
            hitbox.x = size.x / 4 - 8;
            hitbox.y = size.y / 4;
            hitbox.h -= hitbox.y;
        }
    }

    public static class MobSniky extends Mob {

        private static final int HP = 5;
        private static final float SPEED = 10;
        private static final int POINT = 200;

        public MobSniky(RabbitGame game, boolean flipped) {
            super(game.assets.snikyFrames, flipped, HP, SPEED, POINT);
            position.y = 280.0f;
            hitbox.y = hitbox.h / 2;
            hitbox.h = hitbox.h - hitbox.y;
        }
    }
}

File: MiniGame-Python/game/Mob.py
from engine import AnimatedSprite
from random import random

__all__ = ['Mob', 'MobBee', 'MobBlue', 'MobFoxy', 'MobSniky']


class Mob(AnimatedSprite):
    def __init__(self, frames, flipped, hp, speed, points):
        super().__init__(frames, 32, 32)
        self._flipped = flipped
        self._active = True
        self._hp = hp
        self._points = points
        self._speed = speed

        if self._flipped:
            self._position.x = 640.0
            self._velocity.x = -self._speed
        else:
            self._position.x = -32.0
            self._velocity.x = self._speed

    def update(self, dt):
        if self._flipped:
            if self._velocity.x > 0:
                self._velocity.x -= 1.0
                if self._velocity.x < 1.0:
                    self._velocity.x = -self._speed
        else:
            if self._velocity.x < 0:
                self._velocity.x += 1.0
                if self._velocity.x >= 0.0:
                    self._velocity.x = self._speed

        self._position.x += self._velocity.x * dt

        self._animTimer += dt
        if self._animTimer > 0.25:
            self._animTimer = 0.0
            self._animFrame = 1 - self._animFrame

        if self._position.x < -32.0 or self._position.x > 640.0:
            self._active = False

    def gotHit(self, hero):
        if self._hp <= 0:
            return True

        if hero.flipped:
            self._velocity.x = -self._speed * 1.2
        else:
            self._velocity.x = self._speed * 1.2

        return False


class MobBee(Mob):
    def __init__(self, game, flipped):
        super().__init__(game.assets.beeFrames, flipped, hp=2, speed=40, points=20)
        self._position.y = 280.0 - 32 * random() * 32.0


class MobBlue(Mob):
    def __init__(self, game, flipped):
        super().__init__(game.assets.blueFrames, flipped, hp=3, speed=60, points=50)
        self._position.y = 280.0 - 32 * random() * 32.0


class MobFoxy(Mob):
    def __init__(self, game, flipped):
        super().__init__(game.assets.foxyFrames, flipped, hp=2, speed=50, points=100)
        self._position.y = 280.0
        self._hitbox.x = int(self._size.x / 4 - 8)
        self._hitbox.y = int(self._size.y / 4)
        self._hitbox.h -= self._hitbox.y


class MobSniky(Mob):
    def __init__(self, game, flipped):
        super().__init__(game.assets.snikyFrames, flipped, hp=5, speed=10, points=200)
        self._position.y = 280.0
        self._hitbox.y = int(self._size.y / 2)
        self._hitbox.h -= int(self._hitbox.y)
